pgml-dashboard/content/blog/pgml-chat-an-alternative-to-langchain-to-deploy-knowledge-based-chat-agents.md

```diff
@@ -0,0 +1,325 @@
+
+## 4. Connecting to the Real World
+
+Finally, the chatbot needs to be deployed to the real world. This involves:
+
+- Identifying the interface that the users will interact with. This can be Slack, Discord, Teams or your own custom chat platform. Once identified get the API keys for the interface.
+- The chatbot service needs to be hosted on a server that can handle the load of multiple users.
+- The chatbot needs to be integrated with the interface so that it can receive and respond to messages.
+
+# pgml-chat
+`pgml-chat` is a command line tool that allows you to do the following:
+- Build a knowledge base that involves:
+  - Ingesting documents into the database
+  - Chunking documents and storing these chunks in the database
+  - Generating embeddings and storing them in the database
+  - Indexing embeddings for fast query
+- Experimenting with prompts that can be passed to chat completion models like OpenAI's GPT-3 or GPT-4 or Meta's Llama2 models
+- Experimenting with embeddings models that can be used to generate embeddings for the knowledge base
+- Provides a chat interface at command line to evaluate your setup
+- Runs Slack or Discord chat services so that your users can interact with your chatbot.
+
+## Getting Started
+
+Before you begin, make sure you have the following:
+
+- PostgresML Database: Sign up for a free [GPU-powered database](https://postgresml.org/signup)
+- Python version >=3.8
+- OpenAI API key
+
+1. Create a virtual environment and install `pgml-chat` using `pip`:
+
+!!! generic
+
+!!! code_block
+
+```bash
+pip install pgml-chat
+```
+
+!!!
+
+`pgml-chat` will be installed in your virtual environment's PATH.
+
+2. Download `.env.template` file from PostgresML Github repository.
+
+!!! generic
+
+!!! code_block
+
+```bash
+wget https://github.com/postgresml/postgresml/blob/master/pgml-apps/pgml-chat/.env.template 
+```
+
+!!!
+
+3. Copy the template file to `.env`
+
+4. Update environment variables with your OpenAI API key and PostgresML database credentials.
+
+!!! generic
+
+!!! code_block
+
+```bash
+OPENAI_API_KEY=<OPENAI_API_KEY>
+DATABASE_URL=<POSTGRES_DATABASE_URL starts with postgres://>
+MODEL=hkunlp/instructor-xl
+MODEL_PARAMS={"instruction": "Represent the document for retrieval: "}
+QUERY_PARAMS={"instruction": "Represent the question for retrieving supporting documents: "}
+SYSTEM_PROMPT=<> # System prompt used for OpenAI chat completion
+BASE_PROMPT=<> # Base prompt used for OpenAI chat completion for each turn
+SLACK_BOT_TOKEN=<SLACK_BOT_TOKEN> # Slack bot token to run Slack chat service
+SLACK_APP_TOKEN=<SLACK_APP_TOKEN> # Slack app token to run Slack chat service
+DISCORD_BOT_TOKEN=<DISCORD_BOT_TOKEN> # Discord bot token to run Discord chat service
+```
+
+!!!
+
+## Usage
+You can get help on the command line interface by running:
+
+!!! generic
+
+!!! code_block
+
+```bash
```

Comment:
@chillenberger I'm struggling to remember why we need three levels of nesting in markdown, to represent every single code block. Seems like the default style for triple backticks should "handle it".